app/uk/gov/hmrc/taxhistory/connectors/DesNpsConnector.scala

```diff
@@ -44,36 +46,40 @@ class DesNpsConnector @Inject()(val http: HttpGet,
   def getIabds(nino: Nino, year: Int): Future[List[Iabd]] = {
     implicit val hc = basicDesHeaders(HeaderCarrier())
 
-    val timerContext = metrics.startTimer(MetricsEnum.NPS_GET_IABDS)
-
-    val result = http.GET[List[Iabd]](iabdsUrl(nino, year))
-    result.onSuccess { case _ =>
-      timerContext.stop()
-      metrics.incrementSuccessCounter(MetricsEnum.NPS_GET_IABDS)
-    }
-    result.onFailure { case e =>
-      metrics.incrementFailedCounter(MetricsEnum.NPS_GET_IABDS)
-      timerContext.stop()
-      logger.warn(s"DES NPS connector - Error returned from DES NPS connector (getIabds): ${e.toString}: ${e.getMessage}")
+    withMetrics(MetricsEnum.NPS_GET_IABDS) {
+      http.GET[List[Iabd]](iabdsUrl(nino, year))
     }
-    result
   }
 
   def getTaxAccount(nino: Nino, year: Int): Future[NpsTaxAccount] = {
     implicit val hc = basicDesHeaders(HeaderCarrier())
 
-    val timerContext = metrics.startTimer(MetricsEnum.NPS_GET_TAX_ACCOUNT)
+    withMetrics(MetricsEnum.NPS_GET_TAX_ACCOUNT) {
+      http.GET[NpsTaxAccount](taxAccountUrl(nino, year))
+    }
+  }
 
-    val result = http.GET[NpsTaxAccount](taxAccountUrl(nino, year))
-    result.onSuccess { case _ =>
-      timerContext.stop()
-      metrics.incrementSuccessCounter(MetricsEnum.NPS_GET_TAX_ACCOUNT)
+  private def withMetrics[T](metric: MetricsEnum)(codeBlock: => Future[T])(implicit hc: HeaderCarrier) = {
+    val timerContext = metrics.startTimer(metric)
+
+    val result: Future[T] = codeBlock
+
+    result.onSuccess{
```

Comment:
Instead of checking for onSuccess or onFailure on future, I think you could use transform method.

Reply:
Refactored the metrics timing concern into a `ConnectorMetrics` trait that's mixed in to all the connectors now, providing a `withMetrics` thunk.